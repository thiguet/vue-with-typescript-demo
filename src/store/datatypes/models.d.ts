
export enum Measures {
    'UN',
    'L',
    'KG',
    'G',
}

export interface Product {
    name: string;
    measure: keyof typeof Measures;
    qtd: number;
    minQtd: number;
    img?: string;
}
