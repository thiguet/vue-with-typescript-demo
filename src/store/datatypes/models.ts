interface Unique {
    id?: string;
}

export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product extends Unique {
    name: string;
    measure: Measures;
    qtd: number;
    minQtd: number;
    image?: string;
}

export interface User extends Unique {
    name: string;
    email: string;
}

export interface ReportItem {
    text: string;
    value: number;
}

export type ReportData = ReportItem[];

export interface QuantityReportData {
    productName: string;
    productQuantity: number;
}

export type QuantityReport = QuantityReportData[];
export interface MeasuresReportData {
    measureName: string;
    measureQuantity: number;
}

export type MeasuresReport = MeasuresReportData[];

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
    reports = 'reports',
}
