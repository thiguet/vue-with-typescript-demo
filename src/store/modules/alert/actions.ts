import { Actions as VActions, Getters } from 'vuex-smart-module';
import State from './state';
import Mutations, { MutationTypes } from './mutations';

export enum ActionTypes {
    openAlert = 'openAlert',
    closeAlert = 'closeAlert',
}

export default class Actions extends VActions<
    State,
    {} & Getters<State>,
    Mutations,
    Actions
> {
    [ActionTypes.openAlert](payload: string) {
        this.commit(MutationTypes.setMessage, payload);
        this.commit(MutationTypes.openAlert);
    }

    [ActionTypes.closeAlert]() {
        this.commit(MutationTypes.setMessage, '');
        this.commit(MutationTypes.closeAlert);
    }
}
