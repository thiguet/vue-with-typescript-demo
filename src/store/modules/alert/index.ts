import { Module } from 'vuex-smart-module';
import Mutations, { MutationTypes as MTypes } from './mutations';

export const MutationTypes = MTypes;
export const AlertMutations = Mutations;

export class AlertState {
  message!: string;

  display!: boolean;
}

export default new Module({
    state: AlertState,
    mutations: AlertMutations,
});
