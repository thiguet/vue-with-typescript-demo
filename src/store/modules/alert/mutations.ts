import { Mutations as VMutations } from 'vuex-smart-module';
import State from './state';

export enum MutationTypes {
    setMessage = 'setMessage',
    openAlert = 'openAlert',
    closeAlert = 'closeAlert',
}

export default class Mutations extends VMutations<State> {
    [MutationTypes.setMessage](payload: string) {
        this.state.message = payload;
    }

    [MutationTypes.openAlert]() {
        this.state.display = true;
    }

    [MutationTypes.closeAlert]() {
        this.state.display = false;
    }
}
