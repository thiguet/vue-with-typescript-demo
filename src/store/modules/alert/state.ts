export default class AlertState {
    message = '';

    display = false;
}
