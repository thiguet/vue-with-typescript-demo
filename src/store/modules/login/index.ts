import { Module } from 'vuex-smart-module';
import Mutations, { MutationTypes } from './mutations';
import Actions, { ActionTypes } from './actions';
import State from './state';

export { State, Mutations, MutationTypes, Actions, ActionTypes };

export default new Module({
    state: State,
    mutations: Mutations,
    actions: Actions,
});
