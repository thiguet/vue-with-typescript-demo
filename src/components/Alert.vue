<template>
    <div class="alert-bg" v-if="display">
        <div class="alert-card">
            <div class="alert-header">
                <h3 class="title">{{title}}</h3>
            </div>
            <div class="alert-mid">
                <span class="message">{{ message }}</span>
            </div>
            <div class="alert-bottom">
                <button class="close-btn" @click="closeAlert">Fechar</button>
            </div>
        </div>
    </div>
</template>

<script lang="ts">
import { VuexAppModules } from '@/store/datatypes/models';
import { Actions, ActionTypes } from '@/store/modules/alert';

import Vue from 'vue';

import Component from 'vue-class-component';

import { namespace } from 'vuex-class';

const alert = namespace(VuexAppModules.alert);

const { State, Action } = alert;

@Component({
    name: 'Alert',
})
export default class Alert extends Vue {
    @State
    display!: boolean;

    @State
    message!: string;

    @State
    title!: string;

    @Action
    closeAlert!: Actions[ActionTypes.closeAlert];
}
</script>

<style scoped>
.alert-bg {
    display: flex;
    position: absolute;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background: rgba(0, 0, 0, 0.4);
}

.alert-card {
    display: flex;
    position: relative;
    top: 20%;
    left: 15vw;
    width: 70vw;
    height: 40vh;
    flex-direction: column;
    justify-content: center;
    align-content: center;
    background: #fff;
    padding: 20px;
    box-shadow: 1px 1px 10px black;
    border-radius: 5px;
}

.alert-header {
    align-self: center;
}

.alert-bottom {
    margin-top: 20px;
    display: flex;
    flex-direction: row-reverse;
    align-self: center;
}

.alert-bottom > button {
    margin-top: 10px;
}

button {
    padding: 10px;
    cursor: pointer;
    border: none;
    outline-offset: -4px;
    font-family: var(--font-family);
    color: #fff;
    background-color: var(--second-bg);
}

.alert-mid {
    align-self: center;
    height: 35%;
}
.message {
    align-self: center;
    text-align: center;
}
</style>
