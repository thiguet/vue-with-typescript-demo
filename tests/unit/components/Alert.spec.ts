import Vuex, { ActionTree, ModuleTree, MutationTree, Store } from 'vuex';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Alert from '@/components/Alert.vue';
import faker from 'faker';
import { State, ActionTypes } from '@/store/modules/alert';
import { AlertVuex } from '../store/models.d';
import { VuexAppModules } from '@/store/datatypes/models';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Alert', () => {
    let store: Store<State>;
    let alert: AlertVuex;

    const build = () => {
        const wrapper = shallowMount(Alert, {
            store,
            localVue,
        });

        return {
            wrapper,
            message: () => wrapper.find('.message'),
            closeBtn: () => wrapper.find('.close-btn'),
        };
    };

    beforeEach(() => {
        const state: State = {
            display: faker.random.boolean(),
            message: faker.lorem.text(),
        };

        const actions: ActionTree<MutationTree<State>, State> = {
            [ActionTypes.openAlert]: jest.fn(),
            [ActionTypes.closeAlert]: jest.fn(),
        };

        alert = {
            namespaced: true,
            state,
            mutations: {},
            actions,
        };

        const modules: ModuleTree<State> = {
            [VuexAppModules.alert]: {
                ...alert,
            },
        };

        store = new Vuex.Store({
            modules,
        });
    });

    it('renders component', () => {
        const { state } = alert;
        state.message = 'Some default text';
        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('renders main components', () => {
        const { closeBtn, message } = build();

        expect(closeBtn().exists()).toBe(true);
        expect(message().exists()).toBe(true);
    });

    it('message text must be the same as props', () => {
        const { state } = alert;
        const { closeBtn, message } = build();
        state.display = true;

        expect(closeBtn().exists()).toBe(true);
        expect(message().exists()).toBe(true);
    });

    it('calls close fn on click', async () => {
        const { state } = alert;
        state.display = true;

        const { closeBtn } = build();
        const { closeAlert } = alert.actions;

        await closeBtn().trigger('click');

        expect(closeAlert).toHaveBeenCalled();
    });

    it('hides component on isVisible false', () => {
        const { state } = alert;

        state.display = false;

        const { message } = build();

        expect(message().exists()).toBe(false);
    });
});
