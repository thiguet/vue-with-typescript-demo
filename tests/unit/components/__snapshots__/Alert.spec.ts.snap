// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders component 1`] = `
<div class="alert-bg">
  <div class="alert-card">
    <div class="alert-header">
      <h3 class="title"></h3>
    </div>
    <div class="alert-mid"><span class="message">Some default text</span></div>
    <div class="alert-bottom"><button class="close-btn">Fechar</button></div>
  </div>
</div>
`;
